### PR TITLE
Retry read operations up to 10 times on EIO

### PR DESCRIPTION
**Pull request by jviotti**
Some faulty SDCard, or SDCard readers might throw EIO when reading data.
Retrying a couple of times makes the issue go away.

See: https://github.com/resin-io-modules/etcher-image-write/pull/73
Signed-off-by: Juan Cruz Viotti <jviotti@openmailbox.org>

### CODE REVIEW

File: lib/filesystem.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ * @type {Number}
+ */
+const MAXIMUM_RETRIES = 10;
+
+/**
+ * @summary I/O retry base timeout
+ * @constant
+ * @type {Number}
+ */
+const RETRY_BASE_TIMEOUT = 500;
```
Comment (lurch) [inline lib/filesystem.js]: What units is this in? secconds / milliseconds ?

File: lib/filesystem.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ *
+ *   console.log('We wrote a 65536 bytes chunk at position 0');
+ * });
+ */
+exports.writeChunk = (fd, chunk, position, callback, retries = 1) => {
+  fs.write(fd, chunk, 0, chunk.length, position, (error, bytesWritten) => {
+    if (error) {
+
+      // In some faulty SDCards or SDCard readers you might randomly
+      // get EIO errors, but then usually go away after some retries.
```
Comment (lurch) [inline lib/filesystem.js]: then -> they

File: lib/filesystem.js (hunk 1 above)
Comment (jviotti) [inline lib/filesystem.js]: Fixed!

File: lib/filesystem.js (hunk 2 above)
Comment (jviotti) [inline lib/filesystem.js]: Fixed!

File: lib/filesystem.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      // In some faulty SDCards or SDCard readers you might randomly
+      // get EIO errors, but they usually go away after some retries.
+      if (error.code === 'EIO') {
+        if (retries > MAXIMUM_RETRIES) {
+          return callback(error);
+        }
+
+        return setTimeout(() => {
+          return exports.readChunk(fd, size, position, callback, retries + 1);
+        }, RETRY_BASE_TIMEOUT * retries);
```
Comment (lurch) [inline lib/filesystem.js]: Does this increasing-timeout mean that it'll actually take a totally-broken SD card 27.5 seconds before displaying an error? (seems like quite long - I wonder if either `RETRY_BASE_TIMEOUT` or `MAXIMUM_RETRIES` should be reduced?)

**Comment by lurch** [comment]
Should the fact that validation is also now done by chunksize be mentioned in the commit message? (or is that only a minor detail?)

File: lib/filesystem.js (hunk 3 above)
Comment (jviotti) [inline lib/filesystem.js]: Oh, I didn't consider the sum of all of them :/ I'll reduce the timeout to 100ms.

**Comment by jviotti** [comment]
Its a minor internal detail, which is not exposed to the user in any way, so I guess we're fine without a CHANGELOG entry. The high water mark option is also just a "hint" for the input stream, and doesn't mean that it will be strictly respected by the OS.

**Comment by jviotti** [comment]
And now this is suddenly working again, when it previously reliably failed on Appveyor... I'll try on a real Windows setup to see if I can reproduce.

**Comment by jviotti** [comment]
It indeed works now. It might have been an Appveyor issue, or an issue with some of the dependencies. We'll keep an eye on it in case it happens again, but it should be good to merge.

**Comment by jviotti** [comment]
I've done a lot of refactoring (in other commits), rebased, and was able to fix the sporadic invalid file descriptor issue, which seems to have been caused by a race condition on image read stream.

**Pull request merged**